src/main/scala/firrtl/Compiler.scala

```diff
@@ -110,61 +110,110 @@ final case object MidForm extends CircuitForm(1)
+    *   delete annotations
+    * Called by [[runTransform]].
     *
     * @param state Input Firrtl AST
     * @return A transformed Firrtl AST
     */
-  def execute(state: CircuitState): CircuitState
+  protected def execute(state: CircuitState): CircuitState
   /** Convenience method to get annotations relevant to this Transform
     *
     * @param state The [[CircuitState]] form which to extract annotations
     * @return A collection of annotations
     */
   final def getMyAnnotations(state: CircuitState): Seq[Annotation] = state.annotations match {
-    case Some(annotations) => annotations.get(this.getClass)
+    case Some(annotations) => annotations.get(this.getClass) //TODO(azidar): ++ annotations.get(classOf[Transform])
     case None => Nil
   }
-}
+  /** Perform the transform and update annotations.
+    *
+    * @param state Input Firrtl AST
+    * @return A transformed Firrtl AST
+    */
+  final def runTransform(state: CircuitState): CircuitState = {
+    logger.info(s"======== Starting Transform $name ========")
+
+    val (timeMillis, result) = Utils.time { execute(state) }
+
+    logger.info(s"""----------------------------${"-" * name.size}---------\n""")
+    logger.info(f"Time: $timeMillis%.1f ms")
+
+    val remappedAnnotations = propagateAnnotations(state.annotations, result.annotations, result.renames)
 
-trait SimpleRun extends LazyLogging {
-  def runPasses(circuit: Circuit, passSeq: Seq[Pass]): Circuit =
-    passSeq.foldLeft(circuit) { (c: Circuit, pass: Pass) =>
-      val name = pass.name
-      logger.info(s"-------- Starting Pass $name --------")
-      val (timeMillis, x) = Utils.time { pass.run(c) }
-      logger.info(f"Time: $timeMillis%.1f ms")
-      logger.debug(s"Circuit:\n${c.serialize}")
-      logger.info(s"-------- Finished Pass $name --------")
-      x
+    logger.info(s"Form: ${result.form}")
+    logger.debug(s"Annotations:")
+    remappedAnnotations.foreach { a =>
+      logger.debug(a.serialize)
     }
-}
+    logger.debug(s"Circuit:\n${result.circuit.serialize}")
+    logger.info(s"======== Finished Transform $name ========\n")
 
-/** For PassBased Transforms and Emitters
-  *
-  * @note passSeq accepts no arguments
-  * @todo make passes accept CircuitState so annotations can pass data between them
-  */
-trait PassBased extends SimpleRun {
-  def passSeq: Seq[Pass]
-  def runPasses(circuit: Circuit): Circuit = runPasses(circuit, passSeq)
+    CircuitState(result.circuit, result.form, Some(AnnotationMap(remappedAnnotations)), None)
+  }
+
+  /** Propagate annotations and update their names.
+    *
+    * @param inAnno input AnnotationMap
+    * @param resAnno result AnnotationMap
+    * @param renameOpt result RenameMap
+    * @return the updated annotations
+    */
+  final private def propagateAnnotations(inAnno: Option[AnnotationMap], resAnno: Option[AnnotationMap],
+      renameOpt: Option[RenameMap]): Seq[Annotation] = {
```

Comment:
It's not a huge deal but I would prefer:

```scala
final private def propagateAnnotations(inAnno: Option[AnnotationMap],
                                       resAnno: Option[AnnotationMap],
                                       renameOpt: Option[RenameMap]): Seq[Annotation] = {
```
or
```scala
final private def propagateAnnotations(
    inAnno: Option[AnnotationMap], // Note the 4 spaces instead of 2
    resAnno: Option[AnnotationMap],
    renameOpt: Option[RenameMap]): Seq[Annotation] = {
```